### PR TITLE
Lecture 3

### PR DESCRIPTION
**Pull request by RubenVerg**
### Solutions for Lecture 3

cc @chshersh @vrom911

### CODE REVIEW

File: src/Lecture3.hs [hunk 1]
```diff
@@ -156,11 +175,16 @@ monsters, you should get a combined treasure and not just the first
...
 -}
-instance Semigroup (Treasure a) where
-
-
-instance Monoid (Treasure a) where
+instance Semigroup a => Semigroup (Treasure a) where
+    (<>) :: Treasure a -> Treasure a -> Treasure a
+    NoTreasure <> NoTreasure = NoTreasure
+    NoTreasure <> (SomeTreasure t) = SomeTreasure t
+    (SomeTreasure t) <> NoTreasure = SomeTreasure t
```
Comment (chshersh) [inline src/Lecture3.hs:182]: This is an absolutely correct solution!

However, we can notice that these 3 cases could be reduced to 2 with simpler pattern matching!
Let's see what we can unite in here.

If some of the arguments are `NoTreasure`, you actually always return the second one! So you need to inspect only two cases: `NoTreasure` in the first position, and `NoTreasure` in the second one 🙂

File: src/Lecture3.hs [hunk 2]
```diff
@@ -178,7 +202,8 @@ together only different elements.
 Product {getProduct = 6}
 
 -}
-appendDiff3 = error "TODO"
+appendDiff3 :: (Eq a, Semigroup a) => a -> a -> a -> a
+appendDiff3 = foldl1' (<>) . nub .** listOfThree
```
Comment (chshersh) [inline src/Lecture3.hs:206]: Clever solution 😄

File: src/Lecture3.hs [hunk 3]
```diff
@@ -208,10 +233,28 @@ types that can have such an instance.
...
--- instance Foldable Treasure where
+    -- Same again. This time it's a wrapper around two values, but the same principle applies
+
+-- A @List1@ is essentially like @NonEmpty@, so it gets its @Foldable@ instance abusing the fact that it's essentially a list.
+instance Foldable List1 where
+    foldr :: (a -> b -> b) -> b -> List1 a -> b
+    foldr f z (x `List1` xs) = f x $ foldr f z xs
+
+    foldMap :: Monoid m => (a -> m) -> List1 a -> m
+    foldMap f = foldr ((<>) . f) mempty
```
Comment (chshersh) [inline src/Lecture3.hs:248]: That's already the default implementation of `foldMap` so there's not much benefit in implementing it explicitly in exactly the same way 🙂 
But maybe you can think of a different implementation 🤔

File: src/Lecture3.hs [hunk 4]
```diff
@@ -246,4 +294,6 @@ Just [8,9,10]
 [8,20,3]
 
 -}
-apply = error "TODO"
+apply :: Functor f => a -> f (a -> b) -> f b
+-- @flip flap@ is a fun expression, plus, I already have the relude dependency, might as well use it
+apply = flip flap
```
Comment (chshersh) [inline src/Lecture3.hs:299]: The point of the exercise was to implement this function manually and not use one from a library 😅

**Review by chshersh — APPROVED** [review]
Wow, exercises to Lecture 3 are done! 🏆 
You've almost finished the course 🤯 

Keep up the great work and I hope you're enjoying the course so far 🤗 
The finish line is near 🏁

File: src/Lecture3.hs (hunk 4 above)
Comment (RubenVerg) [reply src/Lecture3.hs:299]: well, I implemented it manually first, noticed it wasn't that difficult and then decided to give up to my "don't rewrite code" instinct

**Pull request merged**